### PR TITLE
Subscriptions

### PR DESCRIPTION
**Pull request by Jlrine2**

### CODE REVIEW

File: apps/process-new-granules/process-new-granules-cf.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        Variables:
+          JOBS_TABLE_NAME: !Ref JobsTable
+          USERS_TABLE_NAME: !Ref UsersTable
+          SUBSCRIPTIONS_TABLE_NAME: !Ref SubscriptionsTable
+      Code: src/
+      Handler: process_new_granules.lambda_handler
+      MemorySize: 128
+      Role: !GetAtt Role.Arn
+      Runtime: python3.8
+      Timeout: 30
```
Comment (asjohnston-asf) [inline apps/process-new-granules/process-new-granules-cf.yml:70]: will likely need a (significantly?) longer timeout

Comment (asjohnston-asf) [reply apps/process-new-granules/process-new-granules-cf.yml:70]: and a longer timeout might mandate a less frequent schedule expression

Comment (Jlrine2) [reply apps/process-new-granules/process-new-granules-cf.yml:70]: probably, but for the purposes of testing I'd like to leave it until we have a more robust service

File: apps/process-new-granules/process-new-granules-cf.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+              - Effect: Allow
+                Action:
+                  - logs:CreateLogStream
+                  - logs:PutLogEvents
+                Resource: !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/*"
+              - Effect: Allow
+                Action:
+                  - dynamodb:PutItem
+                  - dynamodb:Query
+                  - dynamodb:GetItem
```
Comment (asjohnston-asf) [inline apps/process-new-granules/process-new-granules-cf.yml:46]:
```suggestion
```

File: apps/process-new-granules/process-new-granules-cf.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                Resource: !Sub "arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${JobsTable}*"
+              - Effect: Allow
+                Action:
+                  - dynamodb:GetItem
```
Comment (asjohnston-asf) [inline apps/process-new-granules/process-new-granules-cf.yml:50]:
```suggestion
                Action: dynamodb:GetItem
```

File: apps/process-new-granules/process-new-granules-cf.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                Resource: !Sub "arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${UsersTable}*"
+              - Effect: Allow
+                Action:
+                  - dynamodb:Scan
```
Comment (asjohnston-asf) [inline apps/process-new-granules/process-new-granules-cf.yml:54]:
```suggestion
                Action: dynamodb:Scan
```

File: apps/process-new-granules/process-new-granules-cf.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    Type: AWS::Lambda::EventInvokeConfig
+    Properties:
+      FunctionName: !Ref Lambda
+      Qualifier: $LATEST
+      MaximumRetryAttempts: 0
+
+  Schedule:
+    Type: AWS::Events::Rule
+    Properties:
+      ScheduleExpression: "rate(1 minute)"
```
Comment (asjohnston-asf) [inline apps/process-new-granules/process-new-granules-cf.yml:82]:
```suggestion
      ScheduleExpression: "rate(1 hour)"
```

File: lib/dynamo/dynamo/subscriptions.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def put_subscription(user, subscription):
+    subscription = {
+        'subscription_id': str(uuid4()),
+        'user_id': user,
+        **subscription
+    }
+
+    validate_subscription(subscription)
```
Comment (asjohnston-asf) [inline lib/dynamo/dynamo/subscriptions.py:36]:
```suggestion
    validate_subscription(subscription)

    subscription = {
        'subscription_id': str(uuid4()),
        'user_id': user,
        **subscription
    }
```

File: apps/api/src/hyp3_api/api-spec/openapi-spec.yml [hunk 7]
```diff
@@ -121,6 +149,34 @@ components:
...
+        - search_parameters
+        - job_specification
+      properties:
+        search_parameters:
+          $ref: "#/components/schemas/search_parameters"
+        job_specification:
+          $ref: "#components/schemas/subscription_job_specification"
+
+    subscriptions_response:
+      description: foo
```
Comment (asjohnston-asf) [inline apps/api/src/hyp3_api/api-spec/openapi-spec.yml]:
```suggestion
    subscription:
```

**Review by asjohnston-asf — CHANGES_REQUESTED** [review]

File: apps/api/src/hyp3_api/api-spec/openapi-spec.yml [hunk 8]
```diff
@@ -121,6 +149,38 @@ components:
...
+        - search_parameters
+        - job_specification
+      properties:
+        subscription_id:
+          $ref: "#/components/schemas/job_or_subscription_id"
+        search_parameters:
+          $ref: "#/components/schemas/search_parameters"
+        job_specification:
+          $ref: "#components/schemas/subscription_job_specification"
+
```
Comment (Jlrine2) [inline apps/api/src/hyp3_api/api-spec/openapi-spec.yml:182]:
```suggestion
          $ref: "#/components/schemas/job_or_subscription_id"
        search_parameters:
          $ref: "#/components/schemas/search_parameters"
        job_specification:
          $ref: "#components/schemas/subscription_job_specification"
        user_id:
          $ref: "#/components/schemas/user_id"
```

**Review by asjohnston-asf — APPROVED** [review]

**Pull request merged**